ipa-core/src/protocol/basics/shard_fin.rs

```diff
@@ -0,0 +1,436 @@
+        inputs: R,
+    ) -> impl Future<Output = Result<R, Error>> + Send
+    where
+        Gate: StepNarrow<S>,
+    {
+        let v = self.dzkp_validator(step, usize::MAX);
+        semi_honest(v.context(), inputs)
+    }
+}
+
+/// This finalizes the MPC execution in sharded context by forwarding the computation results
+/// to the leader shard from all follower shards. Leader shard aggregates them and returns it,
+/// followers set their result to be empty. This implementation only supports semi-honest
+/// security and shouldn't be used directly. Instead [`FinalizerContext`] provides a means
+/// to finalize the execution.
+/// This is a generic implementation that works for both malicious and semi-honest. For the
+/// former, it requires validation phase to be performed after.
+async fn semi_honest<C: ShardedContext, R: ShardAssembledResult<C>>(
+    ctx: C,
+    inputs: R,
+) -> Result<R, crate::error::Error> {
+    if ctx.is_leader() {
+        // leader gets everything from followers
+        let final_r = futures::stream::iter(ctx.peer_shards())
+            .then(|shard| {
+                let stream = ctx.shard_recv_channel::<R::SingleMessage>(shard);
+                R::from_message_stream(stream)
+            })
+            .enumerate()
+            .map(|(i, va)| va.map(|v| (v, i)))
+            .try_fold(inputs, |mut acc, (r, record_id)| {
+                // we merge elements into a single accumulator one by one, thus
+                // record count is indeterminate. A better strategy would be to do
+                // tree-based merge
+                println!("we are in {:?}", ctx.gate());
+                let ctx = ctx.set_total_records(TotalRecords::Indeterminate);
+                async move {
+                    assert_send(acc.merge(ctx, RecordId::from(record_id), r)).await?;
+                    Ok(acc)
+                }
+            })
+            .await?;
+
+        Ok(final_r)
+    } else {
+        // follower just sends its data to the leader
+        let shares = inputs.into_messages();
+        let sz = shares.len();
+        let ctx = ctx.set_total_records(TotalRecords::specified(sz).unwrap_or(TotalRecords::ONE));
+
+        let send_channel = ctx.shard_send_channel::<R::SingleMessage>(ctx.leader());
+
+        seq_join(
+            ctx.active_work(),
+            futures::stream::iter(shares)
+                .enumerate()
+                .map(|(i, v)| send_channel.send(RecordId::from(i), v)),
+        )
+        .try_collect::<()>()
+        .await?;
+
+        send_channel.close(RecordId::from(sz)).await;
+
+        Ok(R::empty())
+    }
+}
+
+/// This type exists to bind [`HV`] and [`B`] together and allow
+/// conversions from [`AdditiveShare<HV>`] to [`BitDecomposed<AdditiveShare<Boolean, B>>`]
+/// and vice versa. Decomposed view is used to perform additions,
+/// share is used to send data to other shards.
+#[derive(Debug, Default)]
+struct Histogram<HV: BooleanArray, const B: usize>
```

Comment:
as an aside, we should probably refactor the `breakdown_reveal` and `apply_dp_noise` to both take and return this type. the fact that one returns `BitDecomposed<AdditiveShare<Boolean, B>>` and the other returns `AdditiveShare<HV>` seems like it could use a touch up.

Reply:
yea, I think the reason for that is we want to keep the inputs to DP noise vectorized for faster additions. DP noise produces final output that does not have to be vectorized, so it uses shares. 

Hybrid would need to do the DP noise addition only on the leader, after this finalization step completes, so it would still probably want bit decomposed view